### PR TITLE
PHP Enums support

### PR DESCRIPTION
**Pull request by pxpm**
### BEFORE - What was wrong? What was happening before this PR?

A powerfull introduction in 8.1 was Enums. Backpack was completely un-ware of them until now. 

It was not possible use a field casted to an `Enum::class`. 

See the following example, a little adaptation of the `Article` model and `ArticleCrudController`:

```php
// the Enum class
enum StatusEnum: string
{
    case DRAFT = 'DRAFT';
    case PUBLISHED = 'PUBLISHED';

    public static function getOptions()
    {
        return array_combine(array_column(self::cases(), 'name'), array_column(self::cases(), 'value'));
    }
}

// In the Article model:

protected $casts = [
        'featured'  => 'boolean',
        'date'      => 'date',
        'status'    => StatusEnum::class
    ];

// In ArticleCrudController - adapting our field to use the enum:

$this->crud->addField([
                'name' => 'status',
                'label' => 'Status',
                'type' => 'select_from_array',
                'options' => StatusEnum::getOptions(),
            ]);

```

### AFTER - What is happening after this PR?

After this PR it works as expected. 

### How did you achieve that, in technical terms?

Basically added a layer between getting and returning the models attributes values. 
That layer will check if Enums is a thing (only available in 8.1), if they are check if the model attribute is any kind of Enum, in case it is, return the appropriate string representation of it. 

**NOTE** For the List/ShowOperation i started from https://github.com/Laravel-Backpack/CRUD/pull/4633 and added support for text and select_from_array columns (probably the ones that make sense to use with an Enum ? )

### Is it a breaking change?

I don't think so, no.

### How can we test the before & after?

You can use the example I gave on this issue description.

### CODE REVIEW

**Comment by ziming** [comment]
Backpack\CRUD\app\Library\CrudPanel\CrudPanel::getAttributeFinalValue(): Argument #1 ($attribute) must be of type string, null given

null is a possible value too :)

**Comment by pxpm** [comment]
> Clicking Edit to show the edit form get me `Backpack\CRUD\app\Library\CrudPanel\CrudPanel::getAttributeFinalValue(): Argument #1 ($attribute) must be of type string, null given`

Should be good now @ziming I will write some tests for it 👍

**Comment by ziming** [comment]
Looks good so far

File: src/app/Library/CrudPanel/Traits/Update.php [hunk 1]
```diff
@@ -163,15 +205,17 @@ private function getModelAttributeValueFromRelationship($model, $field)
 
                 // if `entity` contains a dot here it means developer added a main HasOne/MorphOne relation with dot notation
                 if (Str::contains($field['entity'], '.')) {
-                    return $model->{Str::afterLast($field['entity'], '.')};
+                    dump($this->getAttributeFinalValue($model->{Str::afterLast($field['entity'], '.')}));
```
Comment (ziming) [inline src/app/Library/CrudPanel/Traits/Update.php:208]: remember to remove it :)

**Review by ziming — CHANGES_REQUESTED** [review]

**Comment by pxpm** [comment]
I assigned @jorgetwgroup for testing purposes, and @tabacitu for implementation decisions. 

We can also create a `select_from_enum` column/field that would avoid developer having to setup the `options` manually.

File: tests81/Unit/Models/Enums/StatusEnum.php [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace Backpack\CRUD\Tests81\Unit\Models\Enums;
+
+    enum StatusEnum: string
+    {
+        case DRAFT = 'DRAFT';
+        case PUBLISHED = 'PUBLISHED';
```
Comment (ziming) [inline tests81/Unit/Models/Enums/StatusEnum.php:8]: I think it's better for the "name" and the "value" to be different so troubleshooting will be easier later.

**Review by ziming — CHANGES_REQUESTED** [review]

File: tests81/Unit/Models/Enums/StatusEnum.php [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+        public static function getOptions()
+        {
+            return array_combine(array_column(self::cases(), 'name'), array_column(self::cases(), 'value'));
```
Comment (ziming) [inline tests81/Unit/Models/Enums/StatusEnum.php:12]:
```suggestion
            return array_column(self::cases(), 'value', 'name');
```

The resulting array will be

```php
[
    'NAME_1' => 'Name 1 Value',
    'NAME_2' => 'Name 2 Value',
]
```

**Review by ziming — CHANGES_REQUESTED** [review]

**Review by tabacitu — CHANGES_REQUESTED** [review]
Pedro an I have decided to add enum support a different way - by creating `enum` field and column (actually _adding_ functionality to the enum field but you know what I mean).

That should make this change super-non-breaking, with zero touches to the core 💪

File: tests81/Unit/Models/Enums/StatusEnum.php (hunk 2 above)
Comment (pxpm) [reply tests81/Unit/Models/Enums/StatusEnum.php]: Thanks, I did it! 👍

**Comment by ziming** [comment]
there is already a database enum field in backpack so are you all going to call it php enum to differentiate it?

**Comment by pxpm** [comment]
> there is already a database enum field in backpack so are you all going to call it php enum to differentiate it?
> 
> edit: ah i see so now the enum field support both database and php enum

No zimming, the enum field should work for both enum types (PHP and database) I already did that change in this PR, I am finishing now the enum column too.

**Comment by pxpm** [comment]
@ziming I think it's now finished, tests are passing and I added the column too. 

If you give this a test and found something let me know. 

I am now going to write the docs for it. 

Cheers

File: tests81/Unit/CrudPanel/CrudPanelUpdateTest.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+use Faker\Factory;
+
+/**
+ * @covers Backpack\CRUD\app\Library\CrudPanel\Traits\Update
+ */
+class CrudPanelUpdateTest extends BaseDBCrudPanelTest
+{
+    public function testGetUpdateFieldsWithEnum()
+    {
+        if ((int) app()->version() < 9) {
```
Comment (ziming) [inline tests81/Unit/CrudPanel/CrudPanelUpdateTest.php:15]:
```suggestion
        if ((int) app()->version() < 8) {
```

Enum casting is available in Laravel 8 too

https://laravel.com/docs/8.x/eloquent-mutators#enum-casting

**Review by ziming — CHANGES_REQUESTED** [review]

File: tests81/Unit/CrudPanel/CrudPanelUpdateTest.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            return;
+        }
```
Comment (ziming) [inline tests81/Unit/CrudPanel/CrudPanelUpdateTest.php:17]:
```suggestion
```

PHP Enum casting is available on laravel 8 too

https://laravel.com/docs/8.x/eloquent-mutators#enum-casting

**Review by ziming — CHANGES_REQUESTED** [review]

File: tests81/Unit/CrudPanel/CrudPanelUpdateTest.php (hunk 5 above)
Comment (pxpm) [reply tests81/Unit/CrudPanel/CrudPanelUpdateTest.php:17]: @ziming probably not on all versions because tests were failing stating that `::from()` was not available. 
Tests were failing only on L8+PHP8.1 (this test only runs on PHP8.1+) and we test L9+8.1 and L8+8.1 and was failing on the second scenario with the error I described.

**Comment by pxpm** [comment]
Added docs for field and column https://github.com/Laravel-Backpack/docs/pull/379

File: tests81/Unit/CrudPanel/CrudPanelUpdateTest.php (hunk 5 above)
Comment (pxpm) [reply tests81/Unit/CrudPanel/CrudPanelUpdateTest.php:17]: I've just checked this, and it's only available in https://github.com/laravel/framework/releases/tag/v8.69.0

Backpack only enforces `^8.26`, so if you are on 8.27 it will not complain, but the test will fail. 

We had a reasonable amount of issues that upgrading Laravel constrain would help to fix, is this another one @tabacitu ?

Comment (ziming) [reply tests81/Unit/CrudPanel/CrudPanelUpdateTest.php:17]: I thought tabacitu already said okay to bump in another PR comment previously.

https://github.com/Laravel-Backpack/CRUD/pull/4442#issuecomment-1157608688

I think in general it's okay to use features in newer version of laravel 8. Most laravel packages work this way too.

**Comment by ziming** [comment]
Looking forward to this PR. Hopefully it gets back ported to laravel 8 too

https://github.com/laravel/framework/pull/44445

File: tests81/Unit/CrudPanel/CrudPanelUpdateTest.php (hunk 5 above)
Comment (tabacitu) [reply tests81/Unit/CrudPanel/CrudPanelUpdateTest.php:17]: Yes, I think it's perfect;y reasonable to bump our requirement to `^8.69` - it's the only way we can launch this. Let's do it 💪

**Comment by jorgetwgroup** [comment]
Hello @pxpm thank to you because in this revision everything works as we speak about it.

I will go step by step.

Define the Enum
Screen Shot 2022-10-10 at 20 52 14

Set Model cast
Screen Shot 2022-10-10 at 20 52 21

Add column enum type
Screen Shot 2022-10-10 at 20 52 55

Front
screenshot-backpack test-2022 10 10-20_53_21

Add field "using the same function as column"
Screen Shot 2022-10-10 at 20 53 03

Front
screenshot-backpack test-2022 10 10-20_53_10

On my side everything is working as expected.

About code for me look everything ok, just a warning for @tabacitu this required PHP 8.1 to work normal.

Cheers.

File: tests81/Unit/CrudPanel/CrudPanelUpdateTest.php (hunk 5 above)
Comment (tabacitu) [reply tests81/Unit/CrudPanel/CrudPanelUpdateTest.php:17]: Bumped!

**Comment by tabacitu** [comment]
Excellent work on this one 👏👏👏 Testing and merging it was smooth sailing too. So so happy with it.

slow_clap_cap

Merged! Tagging later today as part of the CRUD 5.4 minor release 🎉
